File: LeetCode/0374. Guess Number Higher or Lower/solution.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
# def guess(num: int) -> int:

"""
25 / 25 test cases passed.
Runtime: 44 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def guessNumber(self, n: int) -> int:
        l, r = 1, n
        while l <= r:
            num = l + (r - l) // 2
            check = guess(num)
            if check == 0:
                return num
            elif check == 1:
                l = num + 1;
            elif check == -1:
                r = num - 1;
        return -1

File: LeetCode/0374. Guess Number Higher or Lower/solution.cpp
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * int guess(int num);
 */
 
/**
 * 48 / 48 test cases passed.
 * Runtime: 68 ms
 * Memory Usage: 50.2 MB 
 */
class Solution {
public:
    int guessNumber(int n) {
        int l = 1, r = n;
        int ans;
        while (l <= r) {
            int num = l + (r - l) / 2;
            int check = guess(num);
            if (!check) {
                ans = num;
                break;
            }
            if (check > 0) l = num + 1;
            else r = num - 1;
        }
        return ans;
    }
};
